elastic/beats
# Fix double fetch

Fixes #35646 by only unpacking project monitors once.

This fixes the ever growing temp folder issue and is more efficient to boot. Previously we would call `fetch` on a monitor source every time it was run, but only cleanup the fetched resource once, when the monitor was unloaded. We now fetch once and cleanup once.

This project also fixes the very confusing issue of two files `browser/project.go` and `browser/source/project.go`, we have renamed `browser/project.go` to `browser/sourcejob.go` which makes reasoning about this change simpler.

## Manual testing

Test with a config like [this one](https://gist.githubusercontent.com/andrewvc/b55c420f688b465352b0abebdfa64e8e/raw/abaf3d0a35b2a92006ce1bbb222d66463c9c4df3/project.yml) by placing it in your `monitors.d` folder in `x-pack/heartbeat`.

Then, in your `heartbeat.yml` enable config reloading. Finally, test by running:


```
mage build && ELASTIC_SYNTHETICS_CAPABLE=true ./heartbeat -e 2>&1 -E cloud.id=$CLOUD_ID -E cloud.auth=$CLOUD_AUTH -E output.elasticsearch.allow_older_versions=true -E logging.level=debug | jq .message | grep "browser project"
```

You should see logs like the following:

```
"browser project: unpack source: /var/folders/pl/wjbh5yrd2dz19mlwpckxrvdm0000gn/T/elastic-synthetics-unzip-1594999618"
"browser project: re-use already unpacked source: /var/folders/pl/wjbh5yrd2dz19mlwpckxrvdm0000gn/T/elastic-synthetics-unzip-1594999618"
"browser project: re-use already unpacked source: /var/folders/pl/wjbh5yrd2dz19mlwpckxrvdm0000gn/T/elastic-synthetics-unzip-1594999618"
// Change made to config file here forcing reload
"browser project: close project source: /var/folders/pl/wjbh5yrd2dz19mlwpckxrvdm0000gn/T/elastic-synthetics-unzip-1594999618"
"browser project: unpack source: /var/folders/pl/wjbh5yrd2dz19mlwpckxrvdm0000gn/T/elastic-synthetics-unzip-2539246959"
"browser project: re-use already unpacked source: /var/folders/pl/wjbh5yrd2dz19mlwpckxrvdm0000gn/T/elastic-synthetics-unzip-2539246959"
```
You should then be able to editor your config in `monitors.d`, perhaps changing the schedule field, etc, forcing it to reload, and see the temp directories used by heartbeat changing, but that only one at a time exists, with the old ones being cleaned up.
